### DISCUSSION TITLE
Node Fetch: Cannot cancel a stream that already has a reader 🙏

### DISCUSSION THREAD

**Discussion by 19Qingfeng** [remix-run/remix#9216] [Q&A] [answered]
I used fetch in Remix, but encountered the following issue：
<img width="668" alt="image" src="https://github.com/remix-run/remix/assets/31119562/6959996a-c9cd-4a49-a707-a15c700f4786">
I'm not sure how I should use fetch on the server to initiate request processing and interrupt requests after timeout.

```tsx
export const loader: LoaderFunction = async ({ context, request }) => {

  const controller = new AbortController()
  fetch('http://google/api.com', {
    signal: controller.signal,
  })
  controller.abort()

  return defer({
  })
}
```
This may seem like a bug, but I do need to retrieve some data through fetch on the server. Is there no other way for Remix to do this?

Can someone help me? 🙏🙏🙏

**Comment by suhaotian** [comment]
What's the point here to abort the request?

**Reply by 19Qingfeng** [reply]
This is just the smallest reproducible example at one end.

**Reply by 19Qingfeng** [reply]
Actually I do have a need for timeout requests to be terminated by controller.abort.

**Reply by suhaotian** [reply]
I think you are looking for https://github.com/axios/axios or https://github.com/suhaotian/xior

```ts
import axios from 'xior';

axios.get('http://google/api.com', {timeout: 10e4})
```

**Reply by 19Qingfeng** [reply]
Actually, the problem lies in the streaming and bundleSize of aixos. We chose to use fetch on our web application site.
Excuse me, does this also mean that fetch cannot be used in Remix at this stage, as I do need to use some res. body functions

**Reply by suhaotian** [reply]
```shell
return promiseRejectedWith (new TypeError\ 'Cannot cancel a stream that already has a reader'));
```

The issue is because your fetch response already solved, then you cancel it, it's impossible.

Try xior, it's built on top of fetch. and fix this situation (when response resolved, will clearTimeout the cancel timer if set timeout)

And xior only ~3kb in gzip(compare to axios 10kb in gzip)

**Reply by 19Qingfeng** [reply]
Thank you for your help, but I'm curious. I just cancelled the fetch request here, but it will cause the entire streaming of Remix to crash, which will cause my application to crash and look like it's remix bug : https://github.com/remix-run/remix/issues/3971

**Reply by suhaotian** [reply]
Yeah, I got it. Thanks!

Remix should try catch the promise reject error and show the error message, now is just crash.

**Comment by suhaotian** [comment, marked as answer]
@19Qingfeng Add this to your `entry.server.tsx`, then it will not crash, the promise reject crash, it's node's default behaviour.

```ts
process.addListener('unhandledRejection', err => {
  console.log('unhandledRejection error: ', err);
  // Do something for example: report log
})
```

**Reply by 19Qingfeng** [reply]
Thank you, this method will damage my application's PM2 related configuration. I found another temporary hack method

```ts
const nodeFetch = require("node-fetch");
let fetchOverlay = false;
export function setupFetch() {
  if (!fetchOverlay) {
    globalThis.fetch = nodeFetch as any;
    fetchOverlay = true;
  }
}
```

I used node fetch to overwrite Remix's built-in fetch on the server

**Reply by suhaotian** [reply]
Since node v18, fetch already support, no need to use node-fetch anymore(expect you need proxy agent feature)

**Reply by 19Qingfeng** [reply]
Remix has implemented a fetch polyfill internally, and I only use node fetch to prevent the aforementioned issues.

**Reply by jacobjohn-transitiv** [reply]
@19Qingfeng , are you executing the setupFetch function within the server.js file? I am also, randomly, getting this rejection error and can't find a good solution for this.
